services/py-api/src/database/repository/participants_repository.py

```diff
@@ -38,15 +39,28 @@ async def delete(
         """
         Deletes the participant which corresponds to the provided object_id
         """
-        try:
-
-            LOG.debug("Deleting participant...", participant_obj_id=obj_id)
 
+        async def db_operation() -> Mapping[str, Any]:
             # According to mongodb docs result is of type _DocumentType:
             # https://pymongo.readthedocs.io/en/4.8.0/api/pymongo/collection.html#pymongo.collection.Collection.find_one_and_delete
             # _id is projected because ObjectID is not serializable.
             # We use the Participant data class to represent the deleted participant.
-            result = await self._collection.find_one_and_delete(filter={"_id": ObjectId(obj_id)}, projection={"_id": 0})
+            delete_result = await self._collection.find_one_and_delete(
+                filter={"_id": ObjectId(obj_id)}, projection={"_id": 0}
+            )
+            # Ignoring mypy type due to mypy err: 'Returning Any from function declared to return "int"  [no-any-return]'
+            # which is not true
+            return delete_result  # type: ignore
+
+        try:
+            LOG.debug("Deleting participant...", participant_obj_id=obj_id)
+
+            # The `TransactionManager.with_transaction` method provides the session and has a built-in retry
+            # mechanism
+            if session:
+                result = await db_operation()
+            else:
+                result = await self._db_manager.retry_db_operation(db_operation, is_read_operation=False)
 
```

Comment:
This approach requires us to remember to always write the retry mechanism in the end and define the operations inside a db_operation(), which could not be the best DEV experience we want to offer and could be error prone. Did you consider writing a decorator for this approach? It seems like it is a perfect candidate to be a decorator, because it includes a nested function that is always used by the retry mechanism. If it is possible, I think this would be a better approach.